atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/pool/CassandraService.java

```diff
@@ -310,9 +314,51 @@ private String getRingViewDescription() {
         return hosts;
     }
 
+    private static double getAverageWithoutSelf(double sum, double self, int count) {
+        return (sum - self) / (count - 1);
+    }
+
+    @VisibleForTesting
+    Set<InetSocketAddress> maybeFilterSlowHosts(Set<InetSocketAddress> desiredHosts) {
```

Comment:
I think that by adding another way to "weight" hosts, we've made it harder to reason about the overall hostselection.  What happens if the slowest host is the only available host?  Especially in combination with the weighted local filtering?  Combining _this_, "local hosts", _and_ active connections, I'm curious if we come up with a combination that is actually worse overall.

Active connections creates a `WeightedHosts` pool - is that something we can potentially leverage here to provide calling functions more options?

There's probably some re-factoring here where at the top level we pass in a list of hosts, and a list of "weighting functions", and our system reasonably attempts to give a "best overall result" when the weighting functions may conflict.

Reply:
`What happens if the slowest host is the only available host?` I've pushed a test case which proves that we do not filter out the host, and we just select it. This makes sense, as the average is calculated without itself, so if its n = 1, the slowness indicator will always be 0, which will always pass. 

Yeah I agree, there is a refactor here which consolidates this logic much more clearly. One argument for my reason why I pushed for _this_ rather than doing a refactor, is that it is stupid simple. In the sense that, the slow hosts are essentially being filtered just like the exclude list logic (i.e. exceptions). So the filter tree goes: Exclude list > slow hosts > Active connections

I also want to highlight the subtle difference here in that, active connections is arguably _preferred_ during execution, while slow hosts are _required_ to be removed (to use Kubernete's terms). Local host filtering is also required as well. Of course, unless the slow host is the only that's left, then we've no choice. 

So the weighting logic comes a bit more complex, as it needs to be some factor of `locality + # of connections + p99`. Which then, we should really test aggressively, as it will be a major change.

Reply:
Another way to see the PR, is that we're simply handling the extreme edge case that `# of active connections` fails to consider, which is that the host may have actual degradation, which is only detectable via latency.

Reply:
Made a prettier version here: https://github.com/palantir/atlasdb/pull/5586